src/IB/FEMechanicsExplicitIntegrator.cpp

```diff
@@ -0,0 +1,739 @@
+                                                             const std::vector<MeshBase*>& meshes,
+                                                             bool register_for_restart,
+                                                             const std::string& restart_read_dirname,
+                                                             unsigned int restart_restore_number)
+    : FEMechanicsBase(object_name, input_db, meshes, register_for_restart, restart_read_dirname, restart_restore_number)
+{
+    commonConstructor(object_name, input_db, meshes, restart_read_dirname, restart_restore_number);
+}
+
+void
+FEMechanicsExplicitIntegrator::preprocessIntegrateData(double current_time, double new_time, int num_cycles)
+{
+    FEMechanicsBase::preprocessIntegrateData(current_time, new_time, num_cycles);
+
+    // Initialize variables.
+    d_X_vecs->copy("solution", { "current", "new", "half" });
+    d_U_vecs->copy("solution", { "current", "new", "half" });
+    d_F_vecs->copy("solution", { "current", "new", "half" });
+    if (d_P_vecs) d_P_vecs->copy("solution", { "current", "new", "half" });
+}
+
+void
+FEMechanicsExplicitIntegrator::postprocessIntegrateData(double current_time, double new_time, int num_cycles)
+{
+    std::vector<std::vector<PetscVector<double>*> > vecs{ d_X_vecs->get("new"),
+                                                          d_U_vecs->get("new"),
+                                                          d_F_vecs->get("new") };
+    if (d_P_vecs) vecs.push_back(d_P_vecs->get("new"));
+    batch_vec_ghost_update(vecs, INSERT_VALUES, SCATTER_FORWARD);
+
+    d_X_vecs->copy("new", { "solution", "current" });
+    d_U_vecs->copy("new", { "solution", "current" });
+    d_F_vecs->copy("new", { "solution", "current" });
+    if (d_P_vecs) d_P_vecs->copy("new", { "solution", "current" });
+
+    FEMechanicsBase::postprocessIntegrateData(current_time, new_time, num_cycles);
+}
+
+void
+FEMechanicsExplicitIntegrator::forwardEulerStep(const double current_time, const double new_time)
+{
+    const double dt = new_time - current_time;
+    for (unsigned int part = 0; part < d_meshes.size(); ++part)
+    {
+        // U^{n+1} := U^{n} + (dt/rho) F^{n}
+        // X^{n+1} := X^{n} + dt       U^{n}
+        computeLagrangianForce(current_time);
+        int ierr;
+        ierr = VecWAXPY(d_U_vecs->get("new", part).vec(),
+                        dt / d_rhos[part],
+                        d_F_vecs->get("current", part).vec(),
+                        d_U_vecs->get("current", part).vec());
+        IBTK_CHKERRQ(ierr);
+        ierr = VecWAXPY(d_X_vecs->get("new", part).vec(),
+                        dt,
+                        d_U_vecs->get("current", part).vec(),
+                        d_X_vecs->get("current", part).vec());
+        IBTK_CHKERRQ(ierr);
+    }
+    // We copy F^{n} into F^{n+1} to make sure that it is stored in the viz
+    // file.
+    d_F_vecs->copy("current", { "new" });
```

Comment:
Will this actually save output? I believe the only vectors libMesh saves are the solutions.

Reply:
The end-time step routine copies new into current and solution.

Reply:
:+1: